lib/sources/astroport/index.ts

```diff
@@ -0,0 +1,208 @@
+    height: number,
+    assetId: string,
+  ) => {
+    const client = await this.getClient();
+    const balance = (
+      await queryContractOnHeight<{ balance: string }>(
+        client,
+        lpContract,
+        height,
+        {
+          balance_at: {
+            address: account,
+            block: height,
+          },
+        },
+      )
+    ).balance;
+    return {
+      address: account,
+      balance,
+      asset: assetId,
+    };
+  };
+
+  getLpContract = async (
+    height: number,
+    pairContract: string,
+  ): Promise<string> => {
+    const client = await this.getClient();
+    const data: any = await queryContractOnHeight(
+      client,
+      pairContract,
+      height,
+      {
+        pair: {},
+      },
+    );
+
+    return data.liquidity_token;
+  };
+
+  getLpExchangeRate = async (
+    height: number,
+    denom: string,
+    lpContract: string,
+  ): Promise<number> => {
+    const client = await this.getClient();
+    const lpTotalSupply = await queryContractOnHeight<string>(
+      client,
+      lpContract,
+      height,
+      {
+        total_supply_at: {
+          block: height,
+        },
+      },
+    );
+
+    const path = '/cosmos.bank.v1beta1.Query/SupplyOf';
+    const request = {
+      denom,
+    };
+    const data = QuerySupplyOfRequest.encode(request).finish();
+    const response = await client.abciQuery({ path, data, height });
+    this.logger.trace('Got response %o', response);
+    if (response.code !== 0) {
+      throw new Error(
+        `Tendermint query error: ${response.log} Code: ${response.code}`,
+      );
+    }
+    const assetTotalSupply = QuerySupplyOfResponse.decode(response.value).amount
+      .amount;
+
+    return Number(assetTotalSupply) / Number(lpTotalSupply);
+  };
+
+  getUsersBalances = async (
+    height: number,
+    multipliers: Record<string, number>,
+    cb: CbOnUserBalances,
+  ): Promise<void> => {
+    for (const [assetId, asset] of Object.entries(this.assets)) {
+      const lpContract = await this.getLpContract(height, asset.pair_contract);
+      const exchangeRate = await this.getLpExchangeRate(
+        height,
+        asset.denom,
+        lpContract,
+      );
+      const multiplier = multipliers[assetId] * exchangeRate;
+
+      const client = await this.getClient();
+      let startAfter = undefined;
+      while (true) {
+        this.logger.info(`Fetching accounts at key ${startAfter}`);
+        const accountsData: { accounts: string[] } =
+          await queryContractOnHeight(client, lpContract, height, {
+            all_accounts: {
+              limit: this.paginationLimit,
+              start_after: startAfter,
+            },
+          });
+        const accounts = accountsData.accounts;
+        if (accounts.length == 0) {
+          break;
+        }
+        startAfter = accounts[accounts.length - 1];
+
+        const withConcurrencyLimit = pLimit(this.concurrencyLimit);
+        let balances = (
```

Comment:
I would call it not `balances` but 'results' or smth to avoid `balances[i].balance`

Reply:
`balances` sounds like array of numbers

Reply:
and it would be nice to specify a type here

Reply:
Alright, I will think how I can restructure it. It was annoying to write and now I guess it is annoying to read as well.

Reply:
no, it is ok, actually :)